### PR TITLE
Implement viewport tracking and reactive rendering in scrollbar

### PR DESCRIPTION
**Pull request by krassowski**
## References

- a step towards https://github.com/jupyterlab/jupyterlab/issues/16332

## Code changes

- `WindowedList.IRenderer.createScrollbarItem` can now return either `HTMLElement` (as before) or `IScrollbarItem` defined as:
  ```ts
  interface IScrollbarItem extends IDisposable {
    /**
     * Render the scrollbar item as an HTML element.
     */
    render: (props: { index: number }) => HTMLElement;
  
    /**
     * Unique item key used for caching.
     */
    key: string;
  }
  ```
- The entire scrollbar is no longer re-rendered on each update of the notebook; this enables adding more reactive information and updating only the cells that have changed; instead `IScrollbarItem.render` is called to modify the node reactively
- a new viewport indicator element is added which shows which cells are visible in the viewport (styling subject to change); this element is independent of cells to avoid scrolling causing re-render of items

## User-facing changes

viewport-indicator

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by srdas** [comment]
@krassowski - This is really nice! As a parallel, I use the mini view in VSCode to traverse quickly to where my edits to the source file are, and I am not sure if this is a useful way to use the viewport in the notebook. If so, adding color to modified cells may be useful. It would definitely be nice to add this and get some feature parity with other IDEs.

**Comment by krassowski** [comment]
Thank you for the suggestion! I implemented it in https://github.com/jupyterlab/jupyterlab/pull/16432.

File: packages/ui-components/src/components/windowedlist.ts [hunk 1]
```diff
@@ -1271,7 +1277,18 @@ export class WindowedList<
...
     if (newWindowIndex !== null) {
-      const [startIndex, stopIndex] = newWindowIndex;
+      const [startIndex, stopIndex, firstVisibleIndex, lastVisibleIndex] =
+        newWindowIndex;
+
+      if (this.scrollbar) {
+        const scrollbarItems = this._renderScrollbar();
+        const first = scrollbarItems[firstVisibleIndex];
+        const last = scrollbarItems[lastVisibleIndex];
+        this._viewportIndicator.style.top = first.offsetTop + 'px';
```
Comment (brichet) [inline packages/ui-components/src/components/windowedlist.ts:1287]:
```suggestion
        this._viewportIndicator.style.top = (first.offsetTop - 1) + 'px';
```

File: packages/ui-components/style/windowedlist.css [hunk 2]
```diff
@@ -38,19 +40,33 @@
...
+  position: relative;
+  z-index: 3;
 }
 
 .jp-WindowedPanel-scrollbar-content > .jp-WindowedPanel-scrollbar-item {
   border-bottom: 1px solid var(--jp-layout-color3);
+  background-color: rgb(from var(--jp-layout-color1) r g b / 50%);
   padding: 2px;
   text-align: left;
+  margin: 2px;
```
Comment (brichet) [inline packages/ui-components/style/windowedlist.css:56]:
```suggestion
  margin: 0 2px;
```

**Review by brichet — COMMENTED** [review]
Thanks @krassowski, this looks great.

There is currently a gap between a selected item and the top separator, but not the bottom separator. And the viewport also seems to be a bit too low



With changes proposed below, it seems more regular:

**Comment by brichet** [comment]
Sorry I split the review for a better readability.

For a long notebook, the virtual scrollbar does not fit in the page, and the scrollbar viewport disappear.



Do you think that the virtual scrollbar could be scrolled to have the scrollbar viewport always displayed ?

**Comment by krassowski** [comment]
>Do you think the virtual scrollbar could be scrolled, to follow the scrollbar viewport ?

Yes, I had the same thought and completely agree that this could be useful, but since this is an additional behaviour change (as of today the virtual scrollbar does not scroll when you move the viewport outside of it) I would defer it to a separate PR. I opened https://github.com/jupyterlab/jupyterlab/issues/16518 to track this.

**Review by brichet — APPROVED** [review]
Thanks @krassowski, looks good to me.

**Pull request merged**